### DISCUSSION TITLE
How to preserve document form field formatting?

### DISCUSSION THREAD

**Discussion by mdmcconnell** [pdfcpu/pdfcpu#992] [Q&A]
I am working with a set of forms that contain many text fields that I fill.  Some of those fields have expectations about character spacing; if characters in the field input have different spacing or width, they look odd.  If I open the documents in acrobat and type input into the fields, it creates non-standard spacing (which looks correct).  Using pdfcpu form.FillForm loses this special character spacing, and the fields look odd.  Can anyone offer hints about where to look for a) what is the formatting specification of the fields in the original pdf; b) a way to affect the formatting of pdfcpu's output fields?

**Comment by hhrutter** [comment]
Can you post a screenshot of an example to get a better idea about what you are talking about?
What do you mean by `expectations about char spacing` normally the filled data is the important and necessary part to move on.

**Comment by mdmcconnell** [comment]
Yes, good idea - not very clear what I am talking about.  Here is the appearance of a field that I type in Acrobat:
![image](https://github.com/user-attachments/assets/29cda1a5-15a9-4926-9a24-b2e87a2b24be)
The letters automatically have additional space, that aligns nicely with lines in the background of the field. 
Here, I try filling the same field with pdfcpu:
![image](https://github.com/user-attachments/assets/a4d36dce-9c04-4409-a9f9-185996a1a188)
It looks like a different font, but more important the character width and spacing does not look aligned.  Somewhere, the field has information about its formatting, but I do not know how to find and preserve that.

**Reply by mdmcconnell** [reply]
When I examine with qpdf --json, it looks like the field is specified as:
`"540 0 R": {
      "/DA": "/HelveticaLTStd-Bold 8.00 Tf 0.000 0.000 0.502 rg",
      "/F": 4,
      "/FT": "/Tx",
      "/Ff": 25165824,
      "/MK": {},
      "/MaxLen": 9,
      "/P": "1922 0 R",
      "/Parent": "534 0 R",
      "/Q": 0,
      "/Rect": [
        469.0,
        689.998,
        576.0,
        703.999
      ],
      "/StructParent": 6,
      "/Subtype": "/Widget",
      "/T": "f1_06[0]",
      "/Type": "/Annot"
    }`

One item these fields share, that is different from other fields with standard character spacing, is the /Ff fieldflags property.  Is this creating the different format?

**Reply by mdmcconnell** [reply]
Possibly, bit 25 field flag is set for these fields, causing the into to fill the entire space allowed by the maxlen.
![image](https://github.com/user-attachments/assets/387f7db6-77bf-4e6b-a5ca-ce9fd8741203)

**Reply by mdmcconnell** [reply]
This the output after processing with pdfdcpu.  The field flags are preserved:
` "540 0 R": {
      "/AP": {
        "/N": "1995 0 R"
      },
      "/DA": "/HelveticaLTStd-Bold 8.00 Tf 0.000 0.000 0.502 rg",
      "/F": 4,
      "/FT": "/Tx",
      "/Ff": 25165824,
      "/MK": {},
      "/MaxLen": 9,
      "/P": "1922 0 R",
      "/Parent": "534 0 R",
      "/Q": 0,
      "/Rect": [
        469.000000000000,
        689.998000000000,
        576.000000000000,
        703.999000000000
      ],
      "/StructParent": 6,
      "/Subtype": "/Widget",
      "/T": "f1_06[0]",
      "/Type": "/Annot",
      "/V": "123456789"
    }`
But there is now an additional /AP field.  Could that be a difference?  Or something inherited from a parent?

**Comment by hhrutter** [comment]
If you have a grid then the natural way to add form fields is so they each fit in a grid cell.
Your PDF wants to make the appearance of the form field content dependent on other page content.
This is a more than exotic usecase and does not align with the intention of a form containing form fields.
Although the PDF spec allows to style your form data there is no way to connect its appearance with other page content.
A form field is just a dumb container for some data rendered on a specific position on a page.

**Reply by mdmcconnell** [reply]
It looks like it is only the comb field flag bit which is creating the formatting, not more information about styling.  And pdfcpu is preserving the comb bit, but something else must be overriding that.  It is interesting that the issue has not been seen before:  it may be an exotic use of PDF features, but this form in some way affects 165mn people every year - it is surely in the list of top 10 most important pdf forms in existence! (Please understand I am being humourous, not making a serious argument!)  I have not been able to identify some obvious difference in JSON representations of the file, so this is beyond my limited understanding of the PDF format currently.
[f1040_acro.pdf](https://github.com/user-attachments/files/17700395/f1040_acro.pdf)
[f1040_pdfcpu.pdf](https://github.com/user-attachments/files/17700396/f1040_pdfcpu.pdf)

**Reply by mdmcconnell** [reply]
Aha - I do see a difference within the files: the appearance in the pdfcpu version sends to a stream:dict object that includes a font resource "152 0 R":
`"obj:1995 0 R": {
        "stream": {
          "dict": {
            "/BBox": [
              0.000000000000,
              0.000000000000,
              107.000000000000,
              14.001000000000
            ],
            "/Filter": "/FlateDecode",
            "/FormType": 1,
            "/Length": 117,
            "/Matrix": [
              1.000000000000,
              0.000000000000,
              0.000000000000,
              1.000000000000,
              0.000000000000,
              0.000000000000
            ],
            "/Resources": {
              "/Font": {
                "/HelveticaLTStd-Bold": "152 0 R"
              }
            },
            "/Subtype": "/Form",
            "/Type": "/XObject"
          }
        }
      }`

It is a simple format object used for all form fields
`"obj:152 0 R": {
        "value": {
          "/BaseFont": "/Helvetica",
          "/Encoding": "/WinAnsiEncoding",
          "/Subtype": "/Type1",
          "/Type": "/Font"
        }
      }`

But, in the Acrobat generated version, the trail of references goes to object 153 0 R, which specificies character widths for every symbol:
`"obj:153 0 R": {
        "value": {
          "/BaseFont": "/HelveticaLTStd-Bold",
          "/Encoding": "/WinAnsiEncoding",
          "/FirstChar": 0,
          "/FontDescriptor": "154 0 R",
          "/LastChar": 255,
          "/Subtype": "/Type1",
          "/Type": "/Font",
          "/Widths": [
            500,
            500,
            500,
            500,
            500, ...`
This same object exists in the pdfcpu output, but is not referenced by any fields.

**Reply by mdmcconnell** [reply]
Something that I cannot see in the un-filled orginal form creates a link between the comb field flag bit and the 153 dictionary object.  Or does pdfcpu generate the comb appearance stream in the same way as Acrobat, but simply "forget" to specifiy it with the field appearance?

**Comment by hhrutter** [comment]
Ha ok I get it... Let me take a closer look and maybe I stand corrected then..

**Comment by hhrutter** [comment]
pdfcpu is creating appearance streams for fields per default, which right now do not take into account the `Comb` and `MaxLen` combo.
You can try setting `NeedAppearances` to true in your configuration and your PDF viewer will take care of the appearance.
This is in the pipeline!

**Comment by mdmcconnell** [comment]
Thank you!  This is already a big improvement!  Is there any downside to using the NeedAppearances flag?  Alignment is not quite perfect, perhaps it has become centre/centre and not left/bottom.
![image](https://github.com/user-attachments/assets/a10833a8-f70e-48c1-b512-94b98ebbd5ba)

**Comment by mdmcconnell** [comment]
One small downside is the viewer will now ask to save changes on closing the file, since it has made alterations.

**Comment by hhrutter** [comment]
pdfcpu renders appearance streams in order to create consistent field appearances across the board, also fontwise.
The only downside is that you may loose that expecially if your input files come from different sources.

**Comment by hhrutter** [comment]
Please do me a favor and create an issue for `combing` support for text fields.
Thank you!

**Comment by hhrutter** [comment]
text combing is supported as of v0.13.0
